pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+
+    Args:
+        encoding_args (dict[str, list[tuple]] or set): Parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``. If a ``set``, all values are set to ``Ellipsis``
+        argnum (list[int]): Numerical indices for arguments
+        qnode (QNode): QNode to infer the ``encoding_args`` and ``argnum`` from
+            if both are ``None``
+    Returns:
+        OrderedDict[str, list[tuple]]: Ordered parameter index dictionary;
+            keys are argument names, values are index tuples for that argument
+            or an ``Ellipsis``
+        list[int]: Numerical indices for arguments
+
+    If ``encoding_args`` are passed, they take precedence over ``argnum``.
+    Passing a set with ``keys`` is an alias for ``{key: ... for key in keys}``.
+    If both, ``encoding_args`` and ``argnum`` are ``None``, all QNode arguments
+    of ``qnode`` that do not have a default value defined are included.
+    Of these arguments, all elements are included in ``encoding_args``
+    """
+    sig_pars = signature(qnode.func).parameters
+    arg_names = list(sig_pars.keys())
+    arg_names_no_def = [name for name, par in sig_pars.items() if par.default is par.empty]
+
+    if encoding_args is None:
+        if argnum is None:
+            encoding_args = OrderedDict((name, ...) for name in arg_names_no_def)
+            argnum = list(range(len(arg_names_no_def)))
+        elif np.isscalar(argnum):
+            encoding_args = OrderedDict({arg_names[argnum]: ...})
+            argnum = [argnum]
+        else:
+            encoding_args = OrderedDict((arg_names[num], ...) for num in argnum)
+    else:
+        requested_names = set(encoding_args)
+        if not all(name in arg_names for name in requested_names):
+            raise ValueError(
+                f"Not all names in {requested_names} are known. " f"Known arguments: {arg_names}"
+            )
+        # Selection of requested argument names from sorted names
+        if isinstance(encoding_args, set):
+            encoding_args = OrderedDict(
+                (name, ...) for name in arg_names if name in requested_names
+            )
+        else:
+            encoding_args = OrderedDict(
+                (name, encoding_args[name]) for name in arg_names if name in requested_names
+            )
+        argnum = [arg_names.index(name) for name in encoding_args]
+
+    return encoding_args, argnum
+
+
+def expand_multi_par(tape, depth=10):
+    """Expand a tape until it does not contain any multi-parameter gates, if possible.
+
+    Args:
+        tape (.QuantumTape): Tape to be expanded
+        depth (int): Maximum expansion depth
+
+    Returns
+        .QuantumTape: Expanded tape
+
+    """
+    if any(len(op.parameters) > 1 for op in tape.operations):
+        stopping_cond = lambda g: (
+            isinstance(g, qml.measure.MeasurementProcess) or len(g.parameters) <= 1
+        )
```

Comment:
Something to keep in mind: multi-parameter gates when the spectrum is not requested _do not need to be expanded_.

Reply:
I think this is non-trivial to support though

Reply:
Exactly, one would have to iteratively compute the Jacobian to find out whether a multi-parameter gate is controlled by a requested parameter. If yes, one would have to decompose and consider the "partial" classical Jacobian and so on...
I don't think this pays off before we know people suffer from the overhead or restrictions due to always decomposing...